src/query/v2/shard_request_manager.hpp

```diff
@@ -172,34 +177,34 @@ class ShardRequestManager : public ShardRequestManagerInterface {
   std::vector<VertexAccessor> Request(ExecutionState<ScanVerticesRequest> &state) override {
```

Comment:
```suggestion
    std::vector<VertexAccessor> Request(ExecutionState<ScanVerticesRequest> &state) override {
      MaybeInitializeExecutionState(state);
      std::vector<ScanVerticesResponse> responses;
  
      SendAllRequests(state);
  
      std::map<Shard, PaginatedResponseState> paginated_response_tracker;
      for (const auto &shard : state.shard_cache) {
        paginated_response_tracker.insert(std::make_pair(shard, PaginatedResponseState::Pending));
      }   
      do {
        AwaitOnPaginatedRequests(state, responses, paginated_response_tracker);
      } while (all_requests_gathered(paginated_response_tracker));
  
      MaybeCompleteState(state);
      // TODO(kostasrim) Before returning start prefetching the batch (this shall be done once we get MgFuture as return
      // result of storage_client.SendReadRequest()).
      return PostProcess(std::move(responses));
    }

```

Reply:
That's it. We use the `PaginatedResponseTracker` to keep track of the pending requests. Look the changes in the `AwaitOnPaginatedRequests`.